### PR TITLE
tests: add test driver implements the particular API being called

### PR DESCRIPTION
**Pull request by maksimmasalski**
Reason is that system call handler for device driver subsystem API
must check that the driver implements the particular API being
called. If it is not implements, will happen kernel oops.

Add test that proves that, for the binded device is impossible to use
sensor_sample_fetch(), will happen error unimplemented system call.

Signed-off-by: Maksim Masalski <maksim.masalski@intel.com>

### CODE REVIEW

**Comment by maksimmasalski** [comment]
@nashif @andrewboie @andyross I'm looking for code review of that PR.

**Comment by jenmwms** [comment]
I’m wondering if we can make this test more comprehensive. As-is it confirms that if we call an api which is not included in the build (impossible to call), then the system will k oops. To me, that could be considered testing the k oops handling an undefined api call. H/e, we really want to test here, if I understand correctly, is ensure that the specified api called is actually getting called (because it is possible) and calling the expected api. IMO We are not fully testing this yet. We should test with the apis included, and we try to call it, then confirm it called that api using some return, etc.

**Comment by maksimmasalski** [comment]
> I’m wondering if we can make this test more comprehensive. As-is it confirms that if we call an api which is not included in the build (impossible to call), then the system will k oops. To me, that could be considered testing the k oops handling an undefined api call. H/e, we really want to test here, if I understand correctly, is ensure that the specified api called is actually getting called (because it is possible) and calling the expected api. IMO We are not fully testing this yet. We should test with the apis included, and we try to call it, then confirm it called that api using some return, etc.

I'm thinking how to make it better.

File: tests/kernel/mem_protect/userspace/src/main.c [hunk 1]
```diff
@@ -1186,6 +1187,31 @@ void test_oops_stackcheck(void)
...
+{
+	int rc;
+
+	expect_fault = true;
+	expected_reason = K_ERR_KERNEL_OOPS;
+
+	struct device *dev = device_get_binding(
+			DT_CHOSEN_ZEPHYR_FLASH_CONTROLLER_LABEL);
+
+	rc = sensor_sample_fetch(dev);
```
Comment (LeiW000) [inline tests/kernel/mem_protect/userspace/src/main.c:1212]: So, will the test case fail?

My understanding is that failure of calling sensor_sample_fetch() is expected result. The test case should be passed when that error message shows up.

Comment (maksimmasalski) [reply tests/kernel/mem_protect/userspace/src/main.c:1212]: Test case will fail, if it is fail, it passed.

**Comment by andrewboie** [comment]
I don't think this is the right way to go about this. You are just showing this works right for ONE API, not all of them, which doesn't do much for us.

If I had to implement this kind of validation I would do the following:

1. For each driver subsystem, verify we have tests that provide coverage of every public API for that driver subsystem. This should already be the case but there may be gaps, but our code coverage reports should show them.
2. For each driver subsystem API that is a system call (they have `__syscall` in the prototype) ensure that the tests for that API are run from a user thread. We definitely have gaps here. I would focus your team's efforts here. There is a lot of work to do in this space. Our code coverage reports don't show whether code is covered from user vs. supervisor mode so this may have to be manually audited.
3. Finally, for each subsystem, implement a dummy driver where ALL the API pointers are NULL. Show that when we run the subsystem test against that dummy driver, that we properly catch this situation.

So to sum up: the best thing we can do right now is to ensure that every driver API that is a system call has code coverage from a user thread.

**Review by nashif — CHANGES_REQUESTED** [review]
need more coverage than one API check.

**Comment by nashif** [comment]
you need to rebase this, this is till trying to use shippable.

**Comment by maksimmasalski** [comment]
> you need to rebase this, this is till trying to use shippable.

Done.

**Comment by maksimmasalski** [comment]
> need more coverage than one API check.

Yes, I aware of it. Andrew explained me exact steps to follow.

**Comment by maksimmasalski** [comment]
Work in Progress. First I made code coverage review of all driver subsystem APIs that are using system call. Found that many of them are not running in user mode in our tests.
